File: src/chapters/numerical-calculus/content/code/differentiation.m
function d = deriv1(f, x, h)
% DERIV1  Central-difference first derivative, error O(h^2).
    if nargin < 3, h = 0.01; end
    d = (f(x + h) - f(x - h)) / (2*h);
end

function d = deriv2(f, x, h)
% DERIV2  Central-difference second derivative, error O(h^2).
    if nargin < 3, h = 0.01; end
    d = (f(x + h) - 2*f(x) + f(x - h)) / h^2;
end

% --- Demo: f(x) = sin(x) at x = 1, h = 0.01 ---
fprintf("f'(1)  ~ %.6f  exact cos(1)  =  %.6f\n", deriv1(@sin, 1, 0.01),  cos(1));
fprintf("f''(1) ~ %.6f  exact -sin(1) = %.6f\n", deriv2(@sin, 1, 0.01), -sin(1));

File: src/chapters/numerical-calculus/content/code/differentiation.rs
// Central-difference first derivative, error O(h^2).
fn deriv1<F: Fn(f64) -> f64>(f: F, x: f64, h: f64) -> f64 {
    (f(x + h) - f(x - h)) / (2.0 * h)
}

// Central-difference second derivative, error O(h^2).
fn deriv2<F: Fn(f64) -> f64>(f: F, x: f64, h: f64) -> f64 {
    (f(x + h) - 2.0 * f(x) + f(x - h)) / (h * h)
}

fn main() {
    let f = |x: f64| x.sin();
    println!("f'(1)  ~ {:.6}  exact cos(1)  = {:.6}", deriv1(f, 1.0, 0.01), 1.0_f64.cos());
    println!("f''(1) ~ {:.6}  exact -sin(1) = {:.6}", deriv2(f, 1.0, 0.01), -1.0_f64.sin());
}
